pkg/storage/filepath/fs.go

```diff
@@ -186,23 +202,41 @@ func (fs *MemoryFS) Write(encoder runtime.Encoder, p string, obj runtime.Object,
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
 
-	versionKey := filepath.Clean(p)
-	if version, ok := fs.versions[versionKey]; ok && version != storageVersion {
+	if rawObj, err := fs.readBuffer(p); err != nil {
+		if os.IsNotExist(err) {
+			// storageVersion == 0 -> this is a create, so it's expected to not exist (continue)
+			// storageVersion != 0 -> object has been deleted, propagate err to avoid a zombie update
+			if storageVersion != 0 {
+				return err
+			}
+		} else {
+			return err
+		}
+	} else if rawObj.version != storageVersion {
+		// this write is outdated
 		return VersionError
+	} else if bytes.Equal(rawObj.data, buf.Bytes()) {
+		// object serialized identically, skip write & version increment
+		return nil
 	}
 
 	dir, err := fs.ensureDir(filepath.Dir(p))
 	if err != nil {
 		return err
 	}
 
-	newVersion, err := resourceVersion(obj)
-	if err != nil {
+	// increment the resource version - it's applied to the object pointer for
+	// the caller in addition to being used to ensure the write is valid
+	newVersion := storageVersion + 1
+	if err := setResourceVersion(obj, newVersion); err != nil {
 		return err
 	}
 
-	dir[filepath.Base(p)] = buf
-	fs.versions[versionKey] = newVersion
+	dir[filepath.Base(p)] = versionedData{
+		version: newVersion,
+		data:    append(make([]byte, 0, len(buf.Bytes())), buf.Bytes()...),
```

Comment:
Thought was to create a new slice of the actual data size since the underlying buffer array might be 2x bigger than needed based on how it grew during encoding.

(I can always swap this back if preferred, I just ended up opportunistically swapping to `[]byte` when adding `versionedData` struct.)

Reply:
hmmm...feels like an unsubstantiated optimization. i've also been wondering if we should use DeepCopyObject() rather than storing as bytes, to save on heap size.

Reply:
I have no idea what's going on with GItHub review comments (I can't respond to the other comment re: this for some reason...?) but the buffer copy is gone as of b3db9ea